serde/src/java/org/apache/hadoop/hive/serde2/JsonSerDe.java

```diff
@@ -63,76 +43,151 @@
+    }
+
+    final String binaryEncodingStr = tbl.getProperty(BINARY_FORMAT, "base64");
+    this.binaryEncoding =
+        BinaryEncoding.valueOf(binaryEncodingStr.toUpperCase());
+
+    this.jsonReader = new HiveJsonReader(this.soi, tsParser);
+    this.jsonWriter = new HiveJsonWriter(this.binaryEncoding, columnNames);
+
+    this.jsonReader.setBinaryEncoding(binaryEncoding);
+    this.jsonReader.enable(HiveJsonReader.Feature.COL_INDEX_PARSING);
 
-    TimestampParser tsParser = new TimestampParser(
-        HiveStringUtils.splitAndUnEscape(tbl.getProperty(serdeConstants.TIMESTAMP_FORMATS)));
-    structReader = new HiveJsonStructReader(rowTypeInfo, tsParser);
-    structReader.setIgnoreUnknownFields(true);
-    structReader.enableHiveColIndexParsing(true);
-    structReader.setWritablesUsage(true);
+    if (writeablePrimitivesDeserialize) {
+      this.jsonReader.enable(HiveJsonReader.Feature.PRIMITIVE_TO_WRITABLE);
+    }
+
+    final String ignoreExtras = tbl.getProperty(IGNORE_EXTRA, "true");
+    if (Boolean.parseBoolean(ignoreExtras)) {
+      this.jsonReader.enable(HiveJsonReader.Feature.IGNORE_UKNOWN_FIELDS);
+    }
+
+    LOG.debug("JSON Struct Reader: {}", jsonReader);
+    LOG.debug("JSON Struct Writer: {}", jsonWriter);
   }
 
   /**
-   * Takes JSON string in Text form, and has to return an object representation above
-   * it that's readable by the corresponding object inspector.
+   * Deserialize an object out of a Writable blob containing JSON text. The
+   * return value of this function will be constant since the function will
+   * reuse the returned object. If the client wants to keep a copy of the
+   * object, the client needs to clone the returned value by calling
+   * ObjectInspectorUtils.getStandardObject().
    *
-   * For this implementation, since we're using the jackson parser, we can construct
-   * our own object implementation, and we use HCatRecord for it
+   * @param blob The Writable (Text) object containing a serialized object
+   * @return A List containing all the values of the row
    */
   @Override
-  public Object deserialize(Writable blob) throws SerDeException {
+  public Object deserialize(final Writable blob) throws SerDeException {
+    final Text t = (Text) blob;
+
+    if (t.getLength() == 0) {
+      if (!this.nullEmptyLines) {
+        throw new SerDeException("Encountered an empty row in the text file");
+      }
+      final int fieldCount = soi.getAllStructFieldRefs().size();
+      return Collections.nCopies(fieldCount, null);
+    }
 
-    Object row;
-    Text t = (Text) blob;
     try {
-      row = structReader.parseStruct(new ByteArrayInputStream((t.getBytes()), 0, t.getLength()));
-      return row;
+      return jsonReader.parseStruct(
+          new ByteArrayInputStream((t.getBytes()), 0, t.getLength()));
     } catch (Exception e) {
-      LOG.warn("Error [{}] parsing json text [{}].", e, t);
+      LOG.warn("Error parsing JSON text [{}].", t, e);
```

Comment:
not sure why this is a warning seems like it is an error ?

Reply:
Thanks for pointing this out.  I am going to lower this logging level to `debug` because there is also an Exception thrown here.  It should either log, or throw, not both.

https://community.oracle.com/docs/DOC-983543#logAndThrow